File: day2/main.py
import re

class PasswordParser:

    def __init__(self, min_: int, max_: int, letter: str, password: str):
        self.min_ = min_
        self.max_ = max_
        self.letter = letter
        self.password = password

    def is_valid(self):
        nbr = self.password.count(self.letter)
        return self.min_ <= nbr <= self.max_
    

# Pattern is min-max caractere: password
def parse_line(line: str):
    pattern = r"(\d+)\-(\d+)\s+(\w){1}:\s+(\w+)"
    matches = re.findall(pattern, line, re.MULTILINE)[0]
    return PasswordParser(int(matches[0]), int(matches[1]), matches[2], matches[3])

def read_file():
    with open('/Users/d.barthelemy/AdventOfCode2020/day2/input.txt', 'r') as f:
        for line in f:
            yield parse_line(line)

def main1():
    lines = read_file()
    count = sum(1 for x in lines if x.is_valid())
    print(f"{count} passwords are valid")

if __name__ == "__main__":
    main1()

File: day2/main.rs
use std::{
    fs::File,
    io::{prelude::*, BufReader}
};

fn read_file(filepath: &str) -> Vec<String> {
    let file = File::open(filepath).expect("no such file");
    let buf = BufReader::new(file);
    buf.lines().map(|l| l.expect("couldn't parse line")).collect()
}

fn parse_line(line: String) -> bool{
    let v: Vec<&str> = line.split(':').map(|c| c.trim()).collect();
    let password = v[1];
    let rules: Vec<&str> = v[0].split(' ').map(|c| c.trim()).collect();
    if rules.len() != 2 {
        return false;
    }
    let bounds: Vec<usize> = rules[0].split('-').map(|x| x.parse::<usize>().unwrap()).collect();
    let letter_count = password.matches(rules[1]).count(); 
    return bounds[0] <= letter_count && letter_count <= bounds[1];
}

fn main() {
    let lines = read_file("input.txt");
    let mut valid_password = 0;
    for line in lines {
        if parse_line(line) {
            valid_password += 1;
        }
    }
    println!("{}", valid_password);
}
